autoload/intero/process.vim

```diff
@@ -376,3 +302,94 @@ function! s:build_complete(job_id, data, event) abort
+    if (executable('intero'))
+        echom 'Intero is installed in your PATH, which may cause problems when using different resolvers.'
+        echom 'This usually happens if you run `stack install intero` instead of `stack build intero`.'
+        echom 'Aborting.'
+        return
+    endif
+
+    if g:intero_use_neomake && !exists(':Neomake')
+        echom 'Neomake not detected. Flychecking will be disabled.'
+    endif
+
+    " Find stack.yaml
+    if (!exists('g:intero_stack_yaml'))
+        " If there's a STACK_YAML environment variable, try to interpret
+        " that.
+        let l:should_cd_to_current_file = empty($STACK_YAML)
+        if l:should_cd_to_current_file
+            " there's no stack yaml env variable, so we can just let stack
+            " figure it out.  Change dir temporarily and see if stack can
+            " find a config
+            silent! lcd %:p:h
+        endif
+
+        " if there's an environment variable, we assume it works
+        " relative to where neovim was started.
+        let l:stack_path_config = systemlist('stack path --config-location')
+        call filter(l:stack_path_config, "v:val =~? '^.*\.yaml'")
+        if empty(l:stack_path_config)
+            echomsg 'Failed to identify a stack.yaml. Does it exist?'
+        else
+            let g:intero_stack_yaml = l:stack_path_config[0]
+        endif
+
+        if l:should_cd_to_current_file
+            silent! lcd -
+        endif
+    endif
+
+    " Ensure that intero is compiled
+    " TODO: Verify that we have a version of intero that the plugin can work with.
+    let l:version = system('stack ' . intero#util#stack_opts() . ' exec --verbosity silent -- intero --version')
+    if v:shell_error
+        let g:intero_built = 0
+        echom 'Intero not installed.'
+        let l:opts = { 'on_exit': function('s:build_complete') }
+        call s:start_compile(10, l:opts)
+    else
+        let g:intero_built = 1
+    endif
+endfunction
+
+function! s:build_terminal_invocation() abort
+    let l:terminal_options = {
+                \ 'on_stdout': function('s:on_stdout'),
+                \ }
+
+    if s:uses_custom_ghci()
+        " Override to use a custom GHCi command.
+        let l:terminal_command = g:intero_ghci_command
+    else
+        " Use the default Intero invocation.
+        let l:terminal_ghci_options = []
+        if exists('g:intero_ghci_options')
+            let l:terminal_ghci_options = ['--ghci-options="' . g:intero_ghci_options .'"']
+        endif
+        let l:terminal_command = join([
+                    \ 'stack',
+                    \ intero#util#stack_opts(),
+                    \ 'ghci',
+                    \ '--with-ghc intero'] +
+                    \ l:terminal_ghci_options +
+                    \ [intero#util#stack_build_opts()]
+                    \ , ' ')
+        let l:terminal_options.cwd = fnamemodify(g:intero_stack_yaml, ':p:h')
+    endif
```

Comment:
Running through Stack and setting the working directory should be done unconditionally, not just on the Intero codepath. Otherwise, this breaks users who have no system installation of GHC.

Reply:
Not sure how it would break?

The idea is to not require Stack at all when you override `g:intero_ghci_command`. Maybe a separate config for Intero's `cwd`?

Reply:
Regardless of if you're using GHCi or Intero, we need it to be in PATH, unless we're expecting the user to set it to an absolute path. Since Intero isn't intended to be installed globally, that looks something like `/home/reuben/myproject/.stack-work/install/x86_64-linux-tinfo6-nopie/lts-9.9/8.0.2/bin/intero`, which means users would need to set it for each project (since it depends on the resolver).

I'm also not certain that just using the path would be enough by itself - I don't know what else Stack does to tell GHCi/Intero which modules are hidden and where to find others.

Reply:
OK, so I'm thinking if you override by using `g:intero_ghci_command`, you really are on your own. If you want to run through Stack, your command has to be something like `stack exec <whatever>`, or `stack repl` more likely. If you aren't using Stack at all, and you have a system-wide installation (Haskell platform or other), you can set it to `ghci` or `cabal repl`.

Reply:
Sounds good to me. The only catch is what we set the working directory to, since we rely on the presence of `stack.yaml` to infer it right now. I think it should be reasonable to just use the path to the current file, but I haven't tested Stack/GHCi to verify it does the right thing here, or how it interacts with module loading.

Reply:
That's right. When overriding to use a custom Stack-based command, you can specify the current working directory yourself with `--stack-yaml`. I can't find any similar thing for `cabal`, so there it depends on Vim's current working directory.

Would perhaps make sense to have a config for `cwd` for the overridden invocation of GHCi, not affecting Intero? We could collect these things in a dict, instead of having `g:intero_ghci_command`, and do something like:

``` vim
g:intero_custom_backend = { 
    \ 'command': 'cabal new-repl', 
    \ 'options': '--whatever', 
    \ 'cwd': '/my/project/path' }
```

Reply:
I like the idea of preemptively using a dict for this sort of configuration options.

Reply:
Yeah, it's starting to make more sense this way. I'll update the PR as soon as I have some time.

Reply:
+1 for dict - if we're going to let the user specify the command, we might as well give them full control over how it's run.

Reply:
Pushed that change now. I called it `g:intero_backend`, so that if we want to represent the default Intero backend in the same way, it could be the default dict value of the same configuration variable.

Also added documentation in my last commit.

Reply:
These changes should've addressed the last remarks.